### PR TITLE
Updated reprojectTexture

### PR DESCRIPTION
**Pull request by slimbuck**
Added the option of limiting reprojectTexture to a single face of a cubemap.

Also moved the options into an option argument as follows. 

The function now has the following signature:
``` javascript
pc.reprojectTexture(source, target, options = { });
```

Which replaces the old style:
```javascript
pc.reprojectTexture(device, source, target, specularPower = 1, numSamples = 1024);
```

### CODE REVIEW

**Comment by mvaligursky** [comment]
- It seems good change, but I would maintain backwards compatibility (worth few lines of code) as I know multiple people are already using this API.
- please list API changes in the description

File: src/graphics/reproject-texture.js
Comment (willeastcott) [inline src/graphics/reproject-texture.js:79]: I don't see how it's possible for `source.device` to be invalid.

Comment (slimbuck) [reply src/graphics/reproject-texture.js:79]: Correct. Question is whether it's necessary to keep supporting user provided device?

**Comment by slimbuck** [comment]
@mvaligursky I've added backwards compatible code, but TBH I had hoped it's not necessary. I'd much rather we don't have code like this littered about the engine. (Though of course I'd also rather we have a better API going forward).

File: src/graphics/reproject-texture.js
Comment (mvaligursky) [reply src/graphics/reproject-texture.js:79]: probably remove options.device completely .. texture always has a valid device.

**Comment by willeastcott** [comment]
This is one of the few cases where I'm 99% sure nobody's using this function (yet).

File: src/graphics/reproject-texture.js
Comment (slimbuck) [reply src/graphics/reproject-texture.js:79]: I'll remove :+1:

File: src/graphics/reproject-texture.js
Comment (mvaligursky) [inline src/graphics/reproject-texture.js:65]: I'm not very keen on this .. it will break in the future when we have base device class and webgl/webgpu implementation. That's a reason why I created https://github.com/playcanvas/engine/pull/3118

Comment (mvaligursky) [reply src/graphics/reproject-texture.js:65]: and it also forces GraphicsDevice to be imported.

Comment (mvaligursky) [reply src/graphics/reproject-texture.js:65]: you might use more correct instanceof in that case

Comment (slimbuck) [reply src/graphics/reproject-texture.js]: yep, done

File: src/graphics/reproject-texture.js
Comment (mvaligursky) [inline src/graphics/reproject-texture.js:81]: I think it'd be better to keep face of the same type (int) .. so assign -1 to it instead of null.

**Review by mvaligursky — APPROVED** [review]
approving with one small comment.

**Comment by mvaligursky** [comment]
also maybe add deprecated warning .. so that later we can remove the support for it.

**Comment by slimbuck** [comment]
@mvaligursky something like this maybe? https://github.com/playcanvas/engine/pull/3167/files#diff-4999255070c170714a51934441b92884e5fbc1fd12351ad47cf5154a75850ed8R73 :+1:

File: src/graphics/reproject-texture.js
Comment (slimbuck) [reply src/graphics/reproject-texture.js:81]: What advantage do you see using the same type? The value 'null' communicates 'no face' better than the value '-1' IMO.

Comment (mvaligursky) [reply src/graphics/reproject-texture.js:81]: it does, but if one day codebase moves to TS, what type do you give to the variable? I believe JIT compilation can optimise better if type does not change.

File: src/graphics/reproject-texture.js
Comment (willeastcott) [inline src/graphics/reproject-texture.js:55]: I think you meant to delete this line?

**Review by willeastcott — APPROVED** [review]

**Comment by willeastcott** [comment]
Approved with one comment.

File: src/graphics/reproject-texture.js
Comment (slimbuck) [reply src/graphics/reproject-texture.js]: yes! thanks

File: src/graphics/reproject-texture.js
Comment (slimbuck) [reply src/graphics/reproject-texture.js:80]: I would do the same (or similar) in TS tbh: https://stackoverflow.com/questions/17220114/how-to-declare-a-type-as-nullable-in-typescript

**Comment by willeastcott** [comment]
Let me make a minor comment about engine architecture. Today, we pass the `GraphicsDevice` into many object constructors (e.g. `Texture`, `VertexBuffer`, `IndexBuffer`, etc). I've talked before about an alternative whereby these classes just hold data/values and the WebGL resources are created/managed in the `GraphicsDevice` and `ForwardRenderer`. This would simplify many classes and remove lots of cyclic dependencies (IMHO). Removing the `GraphicsDevice` parameter from this function is a (small) barrier to us going down that path. It's not enough for me to want to hold up this PR. Just sayin'. 😄

**Comment by slimbuck** [comment]
In that case you'd need the device to perform the rendering?

**Comment by willeastcott** [comment]
Yeah.

**Pull request merged**
